E2E: check log output to detect DATA RACEs

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

We have data races in control plane components which do not shown up in integration or unit tests. With the changes in this PR and the dedicated "kubernetes-e2e-kind-alpha-beta-features-race" job which a) enables building control plane components with race detection (using https://github.com/kubernetes/kubernetes/pull/133834) and b) enables detection of DATA RACE reports on stderr (this PR) it was possible to find:
- https://github.com/kubernetes/kubernetes/issues/135627
- https://github.com/kubernetes/kubernetes/issues/135621
- https://github.com/kubernetes/kubernetes/issues/134371
- https://github.com/kubernetes/kubernetes/issues/134372
- https://github.com/kubernetes/kubernetes/issues/136062
- https://github.com/kubernetes/kubernetes/issues/136179

The e2e.test and ginkgo CLI binary also run with race detection. This found:
- https://github.com/onsi/ginkgo/pull/1624
- https://github.com/kubernetes/kubernetes/issues/134091
 
The failure message is in Markdown format and ready to be copy-and-pasted into a GitHub issue.

#### Which issue(s) this PR is related to:

https://github.com/kubernetes/kubernetes/pull/133834

#### Special notes for your reviewer:

pull-kubernetes-e2e-kind-alpha-beta-features-race was set up to test this PR. Once this PR is merged, a periodic job can be added to prevent future regressions.

Long term the log output checking could also be used for other things ([log spam](https://github.com/kubernetes/kubernetes/issues/109297), [unhandled errors](https://github.com/kubernetes/kubernetes/issues/122005)) but currently that is out of scope.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
E2E: check system logs for DATA RACE reports

While an E2E suite runs, logs of all containers in the "kube-system"
namespace (customizable via command line flag) are retrieved and checked for
"DATA RACE" errors. The kubelet logs are retrieved through the log query
feature. By default, this is attempted on all nodes and it is a test failure
if the log query feature is not enabled. The nodes to check can be customized
via another command line flag.

Data races only get reported if the cluster components were built with data
race detection, so in most clusters this additional checking wouldn't find
anything and therefore is off by default.

The failure message is in Markdown format and ready to be copy-and-pasted into
a GitHub issue.

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

This should build with race detection (I temporarily included https://github.com/kubernetes/kubernetes/pull/133834) and this PR should detect the fake data races...

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

Now with lower log verbosity.

BenTheElder commented:
> The approach for post-run checking is based on @BenTheElder's idea in https://github.com/kubernetes/kubernetes/pull/133394 with a slightly different code setup (separate package).

I've been meaning to send the KEP draft for this and then clean that PR up, was out sick. Working on that now.

I think maybe we want to land the KEP before we get more of these "check after the entire test suite" invariant tests without having formalized what is permitted etc.

> When an E2E test run has finished, logs of all containers in the "kube-system" namespace are retrieved and checked for "DATA RACE" errors. Those only get reported if the cluster components were built with data race detection, so in most clusters this additional checking doesn't report anything.

I'm not sure this approach is reliable, logs may be rotated, etc.

pohly commented on test/e2e/invariants/log.go:
```diff
@@ -0,0 +1,243 @@
+		Name(podName).
+		SubResource("log").
+		Param("container", containerName).
+		Do(ctx).
+		Raw()
```
And then presumably parse the incoming data directly? Yes, that makes sense.

pohly commented:
Perhaps log collection and analysis needs to run in parallel unconditionally and then the results only get reported as failures when the check is enabled?

pohly commented:
It's working. I filed two bugs for the latest findings, using cut-and-paste from the generated report:
- https://github.com/kubernetes/kubernetes/issues/134371
- https://github.com/kubernetes/kubernetes/issues/134372

Providing the source code links is manual work. I'm wondering whether it can be automated... might not be worth it.

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

pohly committed:
e2e podlogs: resume following logs based on last time stamp

If receiving the log output fails for whatever reason, it now uses the last
seen time stamp to continue where it got interrupted.

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

pohly committed:
ginkgo-e2e.sh: support arguments in KUBE_E2E_TEST_ARGS env variable

This is useful when invoked through kind's e2e-k8s.sh which doesn't support
passing through arguments.

The alternatives are:
- Extending e2e-k8s.sh to support passing through arguments.
- Bespoke copies of e2e-k8s.sh which set additional arguments.

Both seem more complex than allowing simple arguments to be passed through the
env variable.

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

pohly committed:
e2e podlogs: support routing log output to some other destination

This enables callers to either do their own file handling or capture the log
output by implementing io.Writer.

While at it, some context handling gets fixed.

BenTheElder commented on test/e2e/logcheck/logcheck.go:
```diff
@@ -0,0 +1,391 @@
+	// defaultNamespacesRE is the default regular expression for which namespaces need to be checked,
+	// if checking is enabled: anything which contains the word "system".
+	//
+	// If we had an API for identifying "system" namespaces, we could use that, but we
```
This also won't get any rotated log files ... I still don't think the tests should be handling logs, the built in logs APIs only give us best effort for recent output, but we can do better than that on the clusters were create for testing, it's just specific to the tools used.

pohly replied:
One more thought on "grab logs after a test run" (no matter how): I tried that initially and found that I had to reduce logging verbosity to avoid getting output discarded. We could try to change the log rotation settings, but that opens another can of worms because we found that they [are load-bearing](https://github.com/kubernetes-sigs/kind/blob/c8778675066f2437874270dc998d1d73d487baa0/hack/ci/e2e-k8s.sh#L184-L189) .

I'd prefer to keep them as similar to other jobs as possible (same verbosity level, same log rotation).

BenTheElder commented:
/cc @aojea @liggitt

BenTheElder commented on test/e2e/logcheck/logcheck.go:
```diff
@@ -0,0 +1,391 @@
+// For DATA RACE, the binaries in the cluster must be compiled with race
+// detection (https://github.com/kubernetes/kubernetes/pull/133834/files).
+// It's also necessary to specify which containers to check. Other output
+// (e.g. kubelet in a kind cluster) currently cannot be checked.
```
Did we check the kubelet system log query feature? 

This feels like a pretty big hack. It's also insecure which isn't the end of the world but ... (e.g. `kind` is not secure, but e2e.test generally doesn't leave a standing pod like this ...)

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

Expected to fail, the job needs to be updated to support the log query feature. I just want to verify that it reports the missing feature in a Ginkgo failure of the log check.

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

With a job fix.

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

pohly commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features-race

pohly commented on hack/tools/golangci-lint/go.mod:
```diff
@@ -226,3 +226,5 @@ require (
 	sigs.k8s.io/yaml v1.4.0 // indirect
 )
+
+replace sigs.k8s.io/logtools => /nvme/gopath/src/sigs.k8s.io/logtools
```
That was accidental, fixed.

pohly commented on staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/roundtrip.go:
```diff
@@ -140,6 +140,7 @@ func roundTripTypes(t *testing.T, scheme *runtime.Scheme, codecFactory runtimese
 			}
 
 			internalGVK := internalVersion.WithKind(kind)
+			t.Logf("starting type %s", internalGVK)
```
Hrmm, I switch between tasks too often. No, doesn't belong here. Removed.

The e2e feature also was a left-over - removed.

BenTheElder commented on test/e2e/logcheck/logcheck.go:
```diff
@@ -0,0 +1,627 @@
+		// How long to wait between queries is a compromise between "too long" (= too much data)
+		// and "too short" (= too much overhead because of frequent queries). It's not clear
+		// where the sweet spot is.
+		ticker := time.NewTicker(300 * time.Second)
```
Unless there's a chance of data loss with low frequency, probably slow polling is preferable to avoid excessive QPS interfering with tests? Even if we don't know the right value, we might note the known trade-offs.

BenTheElder reviewed (commented):
I think this package could do with a short high-level readme to outline the approach, some of the implementation bits are a little surprising

I'm not sure we should be registering flags here, and we should probably be documenting them.

Might be worth temporarily enabling this by default (in a do-not-merge commit) so we can use CI to test that this works as expected with various clusters.

pohly commented:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133844/pull-kubernetes-e2e-gce-csi-serial/2012218005680820224:

```
[FAILED] Unexpected errors during log data collection (see stdout for full log):

    E0116 18:24:26.349034   67683 logcheck.go:473] "PodLogs status" namespace="kube-system" msg="WARNING: pod log: kube-proxy-e2e-50d495fc17-930d0-minion-group-xx0r/kube-proxy: Get \"https://10.40.0.3:10250/containerLogs/kube-system/kube-proxy-e2e-50d495fc17-930d0-minion-group-xx0r/kube-proxy?follow=true&timestamps=true\": dial tcp 10.40.0.3:10250: connect: connection refused\n"
```

If these kind of intermittent errors occur too often in the pull-kubernetes-e2e-kind-alpha-beta-features-race, then some filtering should be able to take care of them to get the noise down.

pohly commented on test/e2e/logcheck/logcheck.go:
```diff
@@ -0,0 +1,627 @@
+	flag.Var(&enabledLogChecks.nodes, "logcheck-nodes-regexp", "all kubelets on nodes matching this regular expressions get checked")
+}
+
+var _ = ginkgo.ReportBeforeSuite(func(ctx ginkgo.SpecContext, report ginkgo.Report) {
```
https://github.com/kubernetes/kubernetes/pull/136274 was merged, so now verify should fail for this PR in its current location:

/test pull-kubernetes-verify

pohly committed:
E2E logcheck: move to "invariants" directory

This check needs closer scrutiny because it hooks into the overall test suite execution.

pohly committed:
E2E logcheck: extend package comments

This explains the overall mode of operation.

pohly committed:
E2E logcheck: don't fail when there are warnings

Warnings should be logged, but don't need to trigger a test failure because
podlogs handles the problem.

While at it, convert text data to string and drop the unnecessary trailing
newline.  Doesn't matter with the current log formatter, but it makes the
intent a bit more obvious.

pohly commented:
/assign @BenTheElder

BenTheElder reviewed (commented):
/lgtm
/approve
/hold

pohly commented:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133844/pull-kubernetes-e2e-kind-alpha-beta-features-race/2013286470181195776

I think these races are covered by https://github.com/kubernetes/kubernetes/issues/136179, fixes pending review.

@BenTheElder: is [the hold](https://github.com/kubernetes/kubernetes/pull/133844#pullrequestreview-3714083939) to get another opinion? From whom?

pohly commented:
/test pull-kubernetes-kind-dra-all-canary

Enables race detection.

pohly commented:
/test pull-kubernetes-kind-dra-all-canary
